### PR TITLE
allow grouping multiple view functions into one elm file

### PR DESCRIPTION
**Pull request by folkertdev**
**my use case** 

I would like to use this library to generate a bunch of html templates (for hakyll). They all depend on 
the same style code, so in most cases, I want to regenerate all my templates when I change some elm code. 

Currently, putting all these view functions (about 10) in a .js file and calling this lib for all of them is too slow (5 seconds), because 10 elm files are generated and compiled (it also seems that my code
as it is imported, is compiled multiple times). 

**proposed fix**

Allow for putting multiple view functions together, compiling them into one elm file. This PR introduces 

```typescript
export interface ViewFunctionConfig {
    viewFunction: string;
    output: string;
    model?: any;
    decoder?: string;
    indent?: number;
    newLines?: boolean;
}

export function grouped(
    rootDir: string, moduleName: string, configs: ViewFunctionConfig[],
    alreadyRun?: boolean, elmMakePath?: string, installMethod?: string): Promise<void[]> 
```

*the moduleName argument is only used to generate a unique hash. Might be better to concatenate the view function names and take that as the hash)*

This function does the writing to file for you, because otherwise the user needs to keep track of the order of the view functions and then reassociate the rendered html with the correct output file later.

Alternatively we could still take the filename as an input but give back a `Promise<[[string, string]]>` where the first element is the rendered html and the other is the file name. 

<hr> 

Does this seem good architecturally? the PR actually contains code that implements this idea but it's not very pretty yet and I'd like to check whether this is the right idea first.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -36,11 +36,11 @@ function parseProjectName(repoName: string): string {
         .replace("/", "$");
 }
 
-function runElmApp(viewHash: string, dirPath: string, model: any): Promise<string> {
+function runElmApp(moduleHash: string, viewHash: string, dirPath: string, model: any): Promise<string> {
 
     return new Promise((resolve, reject) => {
         const Elm = require(path.join(dirPath, "elm.js"));
-        const privateName = `PrivateMain${viewHash}`;
+        const privateName = `PrivateMain${moduleHash}`;
```
Comment (eeue56) [inline index.ts:43]: 💭 why this change?

File: example/index.ts [hunk 2]
```diff
@@ -44,4 +48,18 @@ function runLazyView() {
...
\ No newline at end of file
+runLazyView();
+
+ */
+function runMultiple() { 
+    const configs = 
+        [ { viewFunction: "MyModule.view", model: model
+            , decoder: "MyModule.decodeModel", output: "multiple1.html" } 
+            , { viewFunction: "MyModule.lazyView", model: model
+            , decoder: "MyModule.decodeModel", output: "multiple2.html" }
```
Comment (eeue56) [inline example/index.ts]: 💭 is this the preferred API for you? do you think there could be something better?

File: index.ts (hunk 1 above)
Comment (folkertdev) [reply index.ts:43]: we're combining multiple view functions into one file, so just picking the first view function to generate the hash seems arbitrary. The name would also clash if the same view function is used elsewhere.

File: example/index.ts (hunk 2 above)
Comment (folkertdev) [reply example/index.ts]: I'm not sure, I agree it doesn't feel very nice.

It depends on what one wants to do with the files, I guess. For me, I just want to save them to disk, so the above makes sense. It's possible to return a `Promise<[string]>`, then the user 
has to re-associate the generated html and the file name:

```typescript
function runMultiple() {
    const configs =
        [ { viewFunction: "MyModule.view", model, decoder: "MyModule.decodeModel", output: "multiple1.html" }
        , { viewFunction: "MyModule.lazyView", model, decoder: "MyModule.decodeModel", output: "multiple2.html" }
        ];

    elmStaticHtml.grouped(process.cwd(), "MyModule", configs)
        .then((generatedHtmls) => {
            generatedHtmls
                .map((generatedHtml, i) => fs.writeFileSync(configs[i].output, generatedHtml));
        });
}
```

**Comment by folkertdev** [comment]
I've switched to returning a `Promise<[string]>`. The extra flexibility of having a list of strings
to manipulate is probably worth it. 

To make grouped work together with other elmStaticHtml calls, some further changes were needed.

It seems that elm's port namespace is global, so even though we can compile multiple `main` functions into one javascript file, we can't have multiple ports with the same name. 

There are then two solutions: only generate the ports once (storing them in some `Ports.elm` file, or generating multiple .js files. The second option is simpler and thus the one I went with for now. 
I vaguely remember hearing something about this problem, so maybe this issue is solved in 0.19?

File: example/index.ts [hunk 3]
```diff
@@ -24,24 +25,40 @@ function runTwice() {
...
+function runMultiple() {
+    const configs =
+        [ { viewFunction: "MyModule.view", model, decoder: "MyModule.decodeModel", output: "multiple1.html" }
+        , { viewFunction: "MyModule.lazyView", model, decoder: "MyModule.decodeModel", output: "multiple2.html" }
+        ];
+
+    elmStaticHtml.grouped(process.cwd(), "MyModule", configs)
+        .then((generatedHtmls) => {
+            generatedHtmls
+                .map((generatedHtml, i) => fs.writeFileSync(configs[i].output, generatedHtml));
```
Comment (eeue56) [inline example/index.ts]: Relying on order kinda sucks: why not make make it so that the returned stuff is a list like this:

```
List { generatedHtml : String,  fileOutputName: String }
```
?

File: index.ts [hunk 4]
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
...
+    viewFunction: string;
+    model?: any;
+    decoder?: string;
+    indent?: number;
+    newLines?: boolean;
+}
+
+// compiles multiple view functions into one elm file
+// which is much faster if you're likely to need all of them
+export function grouped(
```
Comment (eeue56) [inline index.ts]: 🎨  the name `grouped` doesn't make much sense to me, though I know why it's called that. How about something like `elmStaticHtml.multiple()` instead?

File: index.ts [hunk 5]
```diff
@@ -36,11 +36,11 @@ function parseProjectName(repoName: string): string {
         .replace("/", "$");
 }
 
-function runElmApp(viewHash: string, dirPath: string, model: any): Promise<string> {
+function runElmApp(moduleHash: string, viewHash: string, dirPath: string, model: any): Promise<string> {
```
Comment (eeue56) [inline index.ts]: 🎨 what's the purpose of the viewHash if we're using a module hash instead? Shouldn't it be a combination of moduleName + viewName?

File: index.ts [hunk 6]
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
...
+    decoder?: string;
+    indent?: number;
+    newLines?: boolean;
+}
+
+// compiles multiple view functions into one elm file
+// which is much faster if you're likely to need all of them
+export function grouped(
+    rootDir: string, moduleName: string, configs: ViewFunctionConfig[],
+    alreadyRun?: boolean, elmMakePath?: string, installMethod?: string): Promise<string[]> {
```
Comment (eeue56) [inline index.ts]: 💭 I wonder if returning `Promise<string, string>[]` would be better: that way it wouldn't block on all of them finishing

File: index.ts [hunk 7]
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
...
+}
+
 export default function elmStaticHtml(rootDir: string, viewFunction: string, options: Options): Promise<string> {
     const viewHash = makeHash(viewFunction);
 
     const dirPath = path.join(rootDir, renderDirName);
 
     if (options.alreadyRun === true) {
-        return runElmApp(viewHash, dirPath, options.model);
+        return runElmApp(viewHash, viewHash, dirPath, options.model);
```
Comment (eeue56) [inline index.ts]: 🤔 Why give the `viewHash` twice?

File: index.ts [hunk 8]
```diff
@@ -82,13 +82,77 @@ function makeHash(viewFunction: string): string {
...
+    indent?: number;
+    newLines?: boolean;
+}
+
+// compiles multiple view functions into one elm file
+// which is much faster if you're likely to need all of them
+export function grouped(
+    rootDir: string, moduleName: string, configs: ViewFunctionConfig[],
+    alreadyRun?: boolean, elmMakePath?: string, installMethod?: string): Promise<string[]> {
+    const moduleHash = makeHash(moduleName);
```
Comment (eeue56) [inline index.ts]: 💭 why does it need a module hash? Why do you need to say which module the functions are in? Don't we already know the module name from the `viewFunction`?

File: index.ts [hunk 9]
```diff
@@ -160,7 +224,45 @@ function runCompiler(viewHash: string,
...
+function runElmAppConfig(moduleHash: string, config: ViewFunctionConfig, rootDir: string): Promise<string> {
+    return runElmApp(moduleHash, makeHash(config.viewFunction), rootDir, config.model);
+}
+
+function runCompilerMany(moduleHash: string,
+                         privateMainPath: string,
+                         rootDir: string, configs: ViewFunctionConfig[], elmMakePath?: string): Promise<string[]> {
+    const options: any = {
+        cwd: rootDir,
+        output: `elm${moduleHash}.js`,
```
Comment (eeue56) [inline index.ts]: 💭 I think this should be a hash of all the viewFunctions joined

**Comment by folkertdev** [comment]
I've started work on the single-port approach we discussed on slack. It makes some of the problems 
with hashes you just highlighted go away and cleans up the code considerably. 

I hope to have that working and incorporate the above comments by saturday. Thanks for the review

**Comment by eeue56** [comment]
Thanks for the PR!

File: index.ts (hunk 6 above)
Comment (folkertdev) [reply index.ts]: Does this have any benefit when the elm app isn't truly parallel? 

You could send in the input via ports, which in theory could give parallelism, but elm doesn't currently support that.  Instead you could spawn the elm app multiple times, but that probably comes with considerable overhead.

But maybe when you have enough models with a small number of view functions, that's worth it?

**Comment by folkertdev** [comment]
found some time today

The above commits

* implement the "one port per generated file" idea.
* remove duplication (`elmStaticHtml` is seens as a singleton `multiple`)
* fix issues with hashes
* thread the `fileOutputName` through

One further idea: compile into separate .js files. This diff gives a 30% speed increase (5.1 to 3.4 seconds) on my machine on `examples/index.ts`. Including all the `PrivateMain` files results in n! compilations, compiling just to the PrivateMain needed is constant.

```diff
diff --git a/index.ts b/index.ts
index 7a3e49f..0a2228f 100644
--- a/index.ts
+++ b/index.ts
@@ -39,7 +39,7 @@ function parseProjectName(repoName: string): string {
 function runElmApp(moduleHash: string, dirPath: string, filenamesAndModels: any[][]): Promise<Output[]> {
 
     return new Promise((resolve, reject) => {
-        const Elm = require(path.join(dirPath, "elm.js"));
+        const Elm = require(path.join(dirPath, `elm${moduleHash}.js`));^M
         const privateName = `PrivateMain${moduleHash}`;
 
         if (Object.keys(Elm).indexOf(privateName) === - 1) {
@@ -218,7 +218,7 @@ function runCompiler(moduleHash: string,
                      configs: ViewFunctionConfig[], elmMakePath?: string): Promise<Output[]> {
     const options: any = {
         cwd: rootDir,
-        output: "elm.js",
+        output: `elm${moduleHash}.js`,^M
         yes: true,
     };
 
@@ -228,7 +228,7 @@ function runCompiler(moduleHash: string,
 
     return new Promise((resolve, reject) => {
         fs.readdir(rootDir, (err, files) => {
-            const actualFiles = files.filter((name) => name.indexOf("PrivateMain") === 0);
+            const actualFiles = files.filter((name) => name.indexOf(`PrivateMain${moduleHash}`) === 0);^M
 
             const compileProcess = compile(actualFiles, options);
             compileProcess.on("exit",

```

This means more files are generated (and the cache removal needs to be a bit more complex) but for 30% speed gain it seems worthwhile. thoughts?

**Comment by eeue56** [comment]
> One further idea: compile into separate .js files. This diff gives a 30% speed increase (5.1 to 3.4 seconds) on my machine on examples/index.ts. Including all the PrivateMain files results in n! compilations, compiling just to the PrivateMain needed is constant.

Why is this faster? 🤔

File: index.ts (hunk 1 above)
Comment (eeue56) [reply index.ts:43]: Gotcha

File: index.ts [hunk 10]
```diff
@@ -82,13 +76,94 @@ function makeHash(viewFunction: string): string {
...
 export default function elmStaticHtml(rootDir: string, viewFunction: string, options: Options): Promise<string> {
     const viewHash = makeHash(viewFunction);
 
+    const config = { decoder: options.decoder
+        , fileOutputName: "placeholder"
+        , indent: options.indent
+        , model: options.model
+        , newLines: options.newLines
+        , viewFunction
+        , viewHash};
```
Comment (eeue56) [inline index.ts]: Not a fan of this style of commas in JS, best to put them afterwards 🎨

File: templates.ts [hunk 11]
```diff
@@ -75,41 +49,134 @@ export function generateRendererFile(
...
+        ${optionsSet}
+    in
+        (decode options) <| ${viewFunction}
+        `;
+}
+
+function generateBody(config: ViewFunctionConfig): string {
+    const optionsSet = generateOptionsSet(config.newLines, config.indent);
+    if (config.decoder) {
+        console.log("withDecoder ->", config.decoder, config.viewFunction);
```
Comment (eeue56) [inline templates.ts]: is this for debugging? if so, make it an optional thing. i used `HIDE_WARNINGS` for this on master

File: templates.ts [hunk 12]
```diff
@@ -75,41 +49,134 @@ export function generateRendererFile(
...
+        configsWithUniqueView
+            .map(generateBody)
+            .join("\n\n");
+
+    const renderersList =
+        configs
+        .map((config) => `render${config.viewHash}`)
+            .join(", ");
+
+    const port = `port htmlOut${hash} : List { generatedHtml : String, fileOutputName: String } -> Cmd msg`;
```
Comment (eeue56) [inline templates.ts:133]: 🎉 much better

File: example/index.ts [hunk 13]
```diff
@@ -24,24 +23,38 @@ function runTwice() {
...
+    const configs =
+        [ { viewFunction: "MyModule.view", model, decoder: "MyModule.decodeModel", fileOutputName: "grouped1.html" }
+        , { viewFunction: "MyModule.lazyView", model, decoder: "MyModule.decodeModel", fileOutputName: "grouped2.html" }
+        , { viewFunction: "MyModule.lazyView", model, decoder: "MyModule.decodeModel", fileOutputName: "grouped3.html" }
+        ];
+
+    elmStaticHtml.multiple(process.cwd(), configs)
+        .then((generatedHtmls) => {
+            generatedHtmls
+                .map((output) => fs.writeFileSync(output.fileOutputName, output.generatedHtml));
```
Comment (eeue56) [inline example/index.ts]: 🎉 much better

File: index.ts [hunk 14]
```diff
@@ -82,13 +76,94 @@ function makeHash(viewFunction: string): string {
...
+    fileOutputName: string;
+    model?: any;
+    decoder?: string;
+    indent?: number;
+    newLines?: boolean;
+}
+
+export interface Output {
+    generatedHtml: string;
+    fileOutputName: string;
```
Comment (eeue56) [inline index.ts:90]: this should be documented in the readme too

File: index.ts [hunk 15]
```diff
@@ -82,13 +76,94 @@ function makeHash(viewFunction: string): string {
...
+}
+
+// compiles multiple view functions into one elm file
+// which is much faster if you're likely to need all of them
+export function multiple(
```
Comment (eeue56) [inline index.ts:95]: usage of this should be documented in the readme

File: example/index.ts [hunk 16]
```diff
@@ -1,20 +1,19 @@
...
-    elmStaticHtml(process.cwd(), "MyModule.view", firstRunOptions)
+    elmStaticHtml.default(process.cwd(), "MyModule.view", firstRunOptions)
     .then((generatedHtml) => {
         fs.writeFileSync("output.html", generatedHtml);
-        elmStaticHtml(process.cwd(), "MyModule.view", secondRunOptions)
+        elmStaticHtml.default(process.cwd(), "MyModule.view", secondRunOptions)
         .then((generatedHtml) => {
             fs.writeFileSync("output2.html", generatedHtml);
-            elmStaticHtml(process.cwd(), "MyModule.view", secondRunOptions)
+            elmStaticHtml.default(process.cwd(), "MyModule.view", secondRunOptions)
```
Comment (eeue56) [inline example/index.ts]: This usage seems wrong,based on the changes you made. Doesn't it return an object now, like the multiple one? I'd rather it didn't return an object if it's a single view

**Comment by folkertdev** [comment]
The compilation is slow because elm files are needlessly compiled multiple times.

say we will generate 3 privatemain files

* the first file is generated, elm.js is created by compiling all PrivateMain*.elm files, which is only one.
* the second file is generated, again elm.js is created by compiling all PrivateMain*.elm files, now two files
* the third file is generated, again elm.js is created by compiling all PrivateMain*.elm files, now 3 files

So 3 elm files results in 1 + 2 + 3 compilations (this is n^2 complexity). Ideally, you'd want to only generate JS once all the elm is generated, but that is impossible to enforce with the singular `elmStaticHtml` commands.

`grouped` works around this problem by only generating one `PrivateMain` for multiple views, but we can solve the problem in general by generating a `elm.js` file per elm file.

**Comment by eeue56** [comment]
Gotcha. I think we should optimize for runtime, rather than compile time. Maybe that can be a different PR later on?

**Comment by folkertdev** [comment]
are multiple applications compiled into one file faster than separate applications? Otherwise there would only be more disk space used, the runtime is unchanged.

but yes this can be a separate thing.

**Comment by eeue56** [comment]
> are multiple applications compiled into one file faster than separate applications? Otherwise there would only be more disk space used, the runtime is unchanged. 

Yep -- requiring a single file in the JS side is faster than requiring as needed

File: example/index.ts (hunk 16 above)
Comment (folkertdev) [reply example/index.ts]: struggling a bit with typescripts module system. 

I think `elmStaticHtml` can't be the default and also be a module. Would it be OK to drop the default and just expose `elmStaticHtml` and `multiple`, and then import as 

```typescript
import { elmStaticHtml, multiple } from "elm-static-html-lib";
```

**Comment by folkertdev** [comment]
* fixed the style issues (console.log use and commas)
* As mentioned above, I think the only way to expose the two functions is to not have a default function, but maybe there is some importing trick that makes it possible that I'm unaware of. 
* Added the new function and types to the readme, Not sure whether the typescript types should be documented as such or as their pure JS versions. I think it's best you make a pass over the readme too to ensure consistent style.

File: example/index.ts [hunk 17]
```diff
@@ -1,12 +1,11 @@
-import elmStaticHtml from "../index";
 import * as fs from "fs";
+import { elmStaticHtml, multiple } from "../index";
```
Comment (eeue56) [inline example/index.ts:2]: 👍  I'm good with this

Comment (eeue56) [reply example/index.ts:2]: Though maybe it could be `single`, but I'm okay with this

**Comment by eeue56** [comment]
Published as 0.1.0! Thanks for your hard work 🎉
